src/raftstore/coprocessor/region_collection.rs

```diff
@@ -0,0 +1,713 @@
+// limitations under the License.
+
+use std::collections::BTreeMap;
+use std::fmt::{Display, Formatter, Result as FmtResult};
+use std::sync::{mpsc, Arc, Mutex};
+use std::usize;
+
+use super::{
+    Coprocessor, CoprocessorHost, ObserverContext, RegionChangeEvent, RegionChangeObserver,
+    RoleObserver,
+};
+use kvproto::metapb::Region;
+use raft::StateRole;
+use raftstore::store::keys::data_end_key;
+use util::collections::HashMap;
+use util::worker::{Builder as WorkerBuilder, Runnable, Scheduler, Worker};
+
+const CHANNEL_BUFFER_SIZE: usize = usize::MAX; // Unbounded
+
+/// `RegionCollection` is used to collect all regions on this TiKV into a collection so that other
+/// parts of TiKV can get region information from it. It registers a observer to raftstore, which
+/// is named `EventSender`, and it simply send some specific types of events through a channel.
+/// In the mean time, `RegionCollectionWorker` keeps fetching messages from the channel, and mutate
+/// the collection according tho the messages. When an accessor method of `RegionCollection` is
+/// called, it also simply send a message to `RegionCollectionWorker`, and the result will be send
+/// back through as soon as it's finished.
+/// In fact, the channel mentioned above is actually a `util::worker::Worker`.
+///
+/// **Caution**: Note that the information in `RegionCollection` is not perfectly precise. Some
+/// regions may be absent while merging or splitting is in progress. Also, `RegionCollection` may
+/// slightly lag actual regions on the TiKV.
+
+/// `RaftStoreEvent` Represents events dispatched from raftstore coprocessor.
+#[derive(Debug)]
+enum RaftStoreEvent {
+    CreateRegion { region: Region },
+    UpdateRegion { region: Region },
+    DestroyRegion { region: Region },
+    RoleChange { region: Region, role: StateRole },
+}
+
+#[derive(Clone, Debug)]
+pub struct RegionInfo {
+    pub region: Region,
+    pub role: StateRole,
+}
+
+impl RegionInfo {
+    pub fn new(region: Region, role: StateRole) -> Self {
+        Self { region, role }
+    }
+}
+
+type RegionsMap = HashMap<u64, RegionInfo>;
+type RegionRangesMap = BTreeMap<Vec<u8>, u64>;
+
+/// `RegionCollection` has its own thread (namely RegionCollectionWorker). Queries and updates are
+/// done by sending commands to the thread.
+enum RegionCollectionMsg {
+    RaftStoreEvent(RaftStoreEvent),
+    /// Get all contents from the collection. Only used for testing.
+    DebugDump(mpsc::Sender<(RegionsMap, RegionRangesMap)>),
+}
+
+impl Display for RegionCollectionMsg {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        match self {
+            RegionCollectionMsg::RaftStoreEvent(e) => write!(f, "RaftStoreEvent({:?})", e),
+            RegionCollectionMsg::DebugDump(_) => write!(f, "DebugDump"),
+        }
+    }
+}
+
+/// `EventSender` implements observer traits. It simply send the events that we are interested in
+/// through the `scheduler`.
+#[derive(Clone)]
+struct EventSender {
```

Comment:
Maybe EventListener is a better name, but it conflict with RocksDB's EventListener.

Reply:
`RegionEventListener`?

Reply:
I think move EventSender into another file is better.

Reply:
It was specifically wrote for RegionCollection, and it only sends events that are useful for RegionCollection. I'm afraid it may be more confusing if we move it into a separated file.